Initial Tableau Reading Support

jdunkerley opened the pull request:
### Pull Request Description

- Adds `Hyper_File` allowing reading a Tableau hyper file.
- Can read the schema and table list.
- Can read the structure of a table.
- Can read data into an Enso Table.

### Important Notes



### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

jdunkerley committed:
Add Tableau

jdunkerley committed:
Building up the bits and bobs.

jdunkerley committed:
Pull hyperd.

jdunkerley committed:
First test of Tableau.

hubertp committed:
Fix integration with HyperAPI

The main problem was that Classloader used during HyperAPI was incorrect
We should be using the one that loaded the class, which was a rather
straightforward change. Additionally we must set correct permissions on
`hyperd`.
sbt will attempt to download necessary dependenciess to
`unmanagedClasspath` if jars are missing.

hubertp committed:
nit

jdunkerley committed:
First functional bits.

jdunkerley committed:
Typo.

jdunkerley committed:
Read structure.

jdunkerley committed:
Read structure fixed.

jdunkerley committed:
Infrastructure for reading a Tableau Table.

jdunkerley committed:
Move inner classes out and split out a bit.

jdunkerley committed:
Fix Nothing handling.

jdunkerley committed:
Split Enso files.
Add ICON and GROUP to Hyper_File.

jdunkerley committed:
Fix dropdowns.

jdunkerley committed:
Add Changelog.

jdunkerley committed:
Add File_Format for Tableau.

jdunkerley committed:
Legal review

jdunkerley committed:
Scala fmt

hubertp committed:
Dependency resolution is a prerequisite for compilation

Should make it more resilient to incremental compilation steps that
appear on CI. Also reproducible locally when `lib`'s contents is removed
manually.

jdunkerley committed:
Add Hyper Copyright

jdunkerley committed:
Useless human...

hubertp committed:
add more info to sbt

radeusgd commented on build.sbt:
```diff
@@ -3606,6 +3617,93 @@ lazy val `std-microsoft` = project
+    fetchZipToUnmanaged := {
+      val unmanagedDirectory = (Compile / unmanagedBase).value
+      if (IO.listFiles(unmanagedDirectory).size < 2) { // Heuristic, should have at least hyperapi jar and os-specific one.
+        System.out.println(
```
Should we use `ManagedLogger`?

hubertp replied:
In general I didn't see the need for any logging. Somehow CI doesn't like my solution and we can't reproduce locally

radeusgd replied:
As you prefer. I think it's useful to see a log for this (at least on debug level) to be able to tell what's happening. I'd especially appreciate a message right before we start a download of the 200Mb file - for fiber connections that's moments, but when you are developing on a train, getting a message explaining to you why the build process just started to block for 5 minutes is actually precious information.

I think it's better to do this via `ManagedLogger` though, as it allows users to manage log levels and gathers logs in one place - print is going to be completely separate from it.

radeusgd commented on distribution/lib/Standard/Tableau/0.0.0-dev/THIRD-PARTY/NOTICE:
```diff
@@ -0,0 +1,12 @@
+'jna-platform', licensed under the Apache-2.0, is distributed with the Tableau.
+The license file can be found at `licenses/APACHE2.0`.
+Copyright notices related to this dependency can be found in the directory `net.java.dev.jna.jna-platform-5.14.0`.
+
```
Forgot to say but I think ideally we should add an entry about `HyperAPI` here. Can be added by modifying `notice-header`.

But current is probably OK too since we do have the `COPYRIGHT-HYPERAPI` file, so feel free to ignore.

radeusgd commented on distribution/lib/Standard/Tableau/0.0.0-dev/src/Hyper_Table.enso:
```diff
@@ -0,0 +1,75 @@
+       ICON metadata
+       Reads The Columns for the Table
+    columns : Vector Hyper_Column
+    columns self =
```
I'd prefer to call this `column_info` or `column_descriptions` or `column_metadata` or something like that, and the same for the type - it's not a (reference) to a column but just metadata description: name, type. One cannot fetch any data just using `Hyper_Column` instance. At least not now. 

So in that regard as you mentioned, `Hyper_Column` is closer to `Column_Description` than to `DB_Column`.

radeusgd commented on distribution/lib/Standard/Tableau/0.0.0-dev/src/Hyper_File.enso:
```diff
@@ -0,0 +1,95 @@
+        array = HyperReader.readSchemas self.file.path
+        Vector.from_polyglot_array array
+
+    ## ICON metadata
+       Returns the name of the current schema.
+       `*` represents all schemas.
+    schema : Text
+    schema self = self.internal_schema
```
What does it mean if `f.schema` returns `"*"`?

jdunkerley replied:
It means accessing all tables - schema isn't used much inside Hyper but it is in theory supported.

radeusgd commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperReader.java:
```diff
@@ -0,0 +1,209 @@
+import org.enso.table.data.table.Column;
+import org.enso.table.problems.ProblemAggregator;
+
+// ** Class responsible for reading from Tableau Hyper files. */
```
```suggestion
/** Class responsible for reading from Tableau Hyper files. */
```

I imagine this was supposed to be a block comment?

radeusgd commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperReader.java:
```diff
@@ -0,0 +1,209 @@
+      }
+    }
+
+    // Check if any files in the hyper directory, otherwise download them.
+    try (var files = Files.list(HYPER_PATH)) {
+      if (files.findAny().isEmpty()) {
+        switch (OSPlatform.CurrentPlatform) {
+          case WINDOWS -> downloadHyper(
+              "https://enso-data-samples.s3.us-west-1.amazonaws.com/tableau/hyperd.exe",
+              "hyperd.exe",
+              false);
+          case MAC_ARM64 -> downloadHyper(
+              "https://enso-data-samples.s3.us-west-1.amazonaws.com/tableau/macos-arm64/hyperd",
+              "hyperd",
+              true);
+          case MAX_X64 -> throw new IOException(
+              "Unsupported platform: Only ARM64 Mac is supported.");
+          case LINUX -> downloadHyper(
+              "https://enso-data-samples.s3.us-west-1.amazonaws.com/tableau/linux/hyperd",
+              "hyperd",
+              true);
+          case OTHER -> throw new IOException(
+              "Unsupported platform: " + OSPlatform.CurrentPlatform);
+        }
+      }
```
Why cannot this executable be distributed together with the library?

jdunkerley replied:
It's massive (>200Mb) so making it fetch it when used.

radeusgd replied:
Okay, fine for now but I think for longer term we should prefer to keep this as part of the library and try using our existing infrastructure to fetch the whole library on demand.

Also that's why I'd like to put this into a `cache` directory - if we ever switch to distributing with library, a cache dir will get more likely to be cleaned up, while this may stay on user's system for longer.

radeusgd commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperReader.java:
```diff
@@ -0,0 +1,209 @@
+
+  private static final Logger LOGGER = Logger.getLogger("enso-hyper-reader");
+
+  private static String getHyperPath() {
+    if (System.getenv("HYPER_PATH") != null) {
+      return System.getenv("HYPER_PATH");
+    }
+    if (System.getenv("ENSO_DATA_DIRECTORY") != null) {
+      return System.getenv("ENSO_DATA_DIRECTORY") + "/hyper";
+    } else {
+      return switch (OSPlatform.CurrentPlatform) {
+        case WINDOWS -> System.getenv("LocalAppData") + "/enso/hyper";
+        case MAC_ARM64, MAX_X64 -> System.getProperty("user.home")
+            + "/Library/Application Support/org.enso/hyper";
+        case LINUX, OTHER -> System.getProperty("user.home") + "/.local/share/enso/hyper";
+      };
+    }
+  }
```
Why `ENSO_DATA_DIRECTORY` and not `ENSO_CACHE_DIRECTORY`?

Also we are duplicating (a minimal portion of) logic from `DistributionManager` - for now probably OK but the more we do it I'm wondering if we should refactor the most basic parts of it to make them re-usable in libs like this.

radeusgd commented on std-bits/tableau/src/main/java/org/enso/tableau/HyperTableColumn.java:
```diff
@@ -0,0 +1,53 @@
+  // ** Type ID for INTERVAL data. */
+  public static final int INTERVAL = 10002;
+
+  static HyperTableColumn FromHyperColumn(int index, TableDefinition.Column hyperColumn) {
```
```suggestion
  static HyperTableColumn fromHyperColumn(int index, TableDefinition.Column hyperColumn) {
```
I think usually our Java naming scheme was that methods start with lowercase, and only types start with uppercase.

radeusgd commented on std-bits/tableau/src/main/java/org/enso/tableau/OSPlatform.java:
```diff
@@ -0,0 +1,30 @@
+  LINUX,
+  OTHER;
+
+  // ** Returns the current platform. */
```
```suggestion
  /** Returns the current platform. */
```

radeusgd commented on std-bits/tableau/src/main/java/org/enso/tableau/TableColumnBuilder.java:
```diff
@@ -0,0 +1,213 @@
+    };
+  }
+
+  // ** Convert a Tableau Interval into either a Duration or a Period (with fallback to String if
+  // needed). */
```
This should be a block comment I suppose.

radeusgd reviewed (approved):
Looks good overall

jdunkerley committed:
Block comments.
Rename method.

jdunkerley committed:
Java Format.

jdunkerley committed:
Add the start of unit tests for Tableau.

jdunkerley committed:
Handle process failure more gracefully.
Handle missing file.

jdunkerley committed:
Java fmt and sample hyper files.

hubertp committed:
sbt now has up-to-date list of unmanaged deps

Previously the task was reporting an empty sequence despite fetching the
required jars.

jdunkerley committed:
Read test part 1.

GregoryTravis commented on build.sbt:
```diff
@@ -3606,6 +3617,92 @@ lazy val `std-microsoft` = project
+        "x86_64"
+      }
+      new URI(
+        s"https://downloads.tableau.com/tssoftware//tableauhyperapi-java-$platform-$arch-release-main.$tableauVersion.zip"
```
```suggestion
        s"https://downloads.tableau.com/tssoftware/tableauhyperapi-java-$platform-$arch-release-main.$tableauVersion.zip"
```

(Or is that there for a reason?)

hubertp replied:
It's not. Thanks!

GregoryTravis commented on distribution/lib/Standard/Tableau/0.0.0-dev/src/Hyper_File.enso:
```diff
@@ -0,0 +1,100 @@
+    @schema (hyper -> make_schema_selector hyper True)
+    set_schema : Text -> Hyper_File
+    set_schema self schema =
+        if schema == self.schema then self else
```
Is `schema` a collection of tables, or is it table metadata?

hubertp reviewed (approved):
Approving my own sbt changes because I'm always right

![image](https://github.com/user-attachments/assets/f41a4221-71bf-4aaf-80d7-d7da3c9b28cd)

jdunkerley committed:
Fix failing test.

jdunkerley committed:
Fix failing test.

jdunkerley committed:
Java format.

hubertp committed:
Ensure `unmanagedJars` fetches jars

Since our CI builds things multiple times and it might happen
`unmanagedClasspath` was not triggered before `unmanagedJars`, we could
have a situation when necessary jars were not copied to the
distribution.
This change ensures that is no longer the case.

hubertp committed:
Update build.sbt

Co-authored-by: GregoryTravis <greg.m.travis@gmail.com>

radeusgd committed:
Merge branch 'develop' into wip/jd/tableau
